internal/core/src/mmap/ChunkVector.h

```diff
@@ -0,0 +1,197 @@
+            chunk_id < this->counter_ && chunk_offset < chunk.size(),
+            fmt::format("index out of range, index={}, chunk_offset={}, cap={}",
+                        chunk_id,
+                        chunk_offset,
+                        chunk.size()));
+        if constexpr (IsMmap) {
+            return chunk.get(chunk_offset);
+        } else {
+            return chunk[chunk_offset];
+        }
+    }
+
+    ChunkViewType<Type>
+    view_element(int64_t chunk_id, int64_t chunk_offset) {
+        std::shared_lock lck(mutex_);
+        auto chunk = vec_[chunk_id];
+        if constexpr (IsMmap) {
+            return chunk.view(chunk_offset);
+        } else if constexpr (std::is_same_v<std::string, Type>) {
+            return std::string_view(chunk[chunk_offset].data(),
+                                    chunk[chunk_offset].size());
+        } else if constexpr (std::is_same_v<Array, Type>) {
+            auto& src = chunk[chunk_offset];
+            return ArrayView(const_cast<char*>(src.data()),
+                             src.byte_size(),
+                             src.get_element_type(),
+                             src.get_offsets_in_copy());
+        } else {
+            return chunk[chunk_offset];
+        }
+    }
+
+    void*
+    get_chunk_data(int64_t index) override {
+        AssertInfo(index < this->counter_,
+                   fmt::format("index out of range, index={}, counter_={}",
+                               index,
+                               this->counter_));
+        std::shared_lock lck(mutex_);
+        return vec_[index].data();
+    }
+
+    int64_t
+    get_chunk_size(int64_t index) override {
+        AssertInfo(index < this->counter_,
+                   fmt::format("index out of range, index={}, counter_={}",
+                               index,
+                               this->counter_));
+        std::shared_lock lck(mutex_);
+        return vec_[index].size();
+    }
+
+    void
+    clear() override {
+        std::lock_guard lck(mutex_);
+        this->counter_ = 0;
+        vec_.clear();
+    }
+
+ private:
+    mutable std::shared_mutex mutex_;
+    storage::MmapChunkDescriptor mmap_descriptor_ = std::nullopt;
+    std::deque<ChunkImpl> vec_;
+};
+
+template <typename Type>
+ChunkVectorPtr<Type>
+SelectChunkVectorPtr(storage::MmapChunkDescriptor& mmap_descriptor) {
+    if constexpr (!IsVariableType<Type>) {
+        if (mmap_descriptor.has_value()) {
+            return std::make_unique<
+                ThreadSafeChunkVector<Type, FixedLengthChunk<Type>, true>>(
+                mmap_descriptor);
+        } else {
+            return std::make_unique<ThreadSafeChunkVector<Type>>();
+        }
+    } else if constexpr (IsVariableTypeSupportInChunk<Type>) {
+        if (mmap_descriptor.has_value()) {
+            return std::make_unique<
+                ThreadSafeChunkVector<Type, VariableLengthChunk<Type>, true>>(
+                mmap_descriptor);
+        } else {
```

Comment:
If optional check is only used here, the call stack under this don't have to accept an optional.

Reply:
use mmap_descriptor  use optional can reduce some of the conditional judgments for creating concurrent vector in InsertRecord.